File: Leetcode/Leetcode-DSA-course/Trees-and-Graphs/103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ans = []
        if not root:
            return ans
        queue = deque([root])
        level = 0
        while queue:
            length = len(queue)
            currList = []
            for _ in range(length):
                node = queue.popleft()
                currList.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            if level & 1:
                ans.append(currList[::-1])
            else:
                ans.append(currList)
            level += 1
        return ans

File: Leetcode/Leetcode-DSA-course/Trees-and-Graphs/103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    static void reverseNums(vector<int>& nums) {
        int i = 0, j = nums.size() - 1;
        while (i < j)
            swap(nums[i++], nums[j--]);
    }
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if (!root)
            return {};
        queue<TreeNode*> queue;
        queue.push(root);
        int level = 0;
        while (!queue.empty()) {
            int length = queue.size();
            vector<int> currLevel;
            for (int i = 0; i < length; i++) {
                TreeNode* node = queue.front();
                queue.pop();
                currLevel.push_back(node->val);
                if (node->left)
                    queue.push(node->left);
                if (node->right)
                    queue.push(node->right);
            }
            if (level & 1) {
                reverseNums(currLevel);
            }
            ans.push_back(currLevel);
            level++;
        }
        return ans;
    }
};